### PR TITLE
[BIOMAGE-1746] - Add empty cellset notification

### PR DESCRIPTION
**Pull request by aerlaut**
# Description
This PR adds a notification to show to the user if cell set creation by expression fails because the created cell set is empty. 

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1746

#### Link to staging deployment URL (or set N/A)
https://ui-agi-ui642.scp-staging.biomage.net/

#### Links to any PRs or resources related to this PR
N/A

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [X] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [X] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [X] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [X] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [X] Relevant Github READMEs updated **or** no GitHub README updates required.
- [X] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: src/redux/actions/cellSets/createCellSetByExpression.js [hunk 1]
```diff
@@ -10,14 +8,8 @@ const createCellSetByExpression = (
...
   };
 
-  try {
-    const result = await fetchWork(experimentId, body, getState, { broadcast: true });
-    return result;
-  } catch (e) {
-    console.error(e);
-    pushNotificationMessage('error', endUserMessages.ERROR_FETCHING_CELL_SETS);
-    throw new Error(e);
-  }
```
Comment (ivababukova) [inline src/redux/actions/cellSets/createCellSetByExpression.js]: why are you removing all the error handling in here?

File: src/utils/experimentUpdatesHandler.js [hunk 2]
```diff
@@ -63,16 +65,43 @@ const onQCUpdate = (update, dispatch, experimentId) => {
 };
 
 const onGEM2SUpdate = (update, dispatch, experimentId) => {
-  const processingConfig = update?.item?.processingConfig;
+  const { response: { error, errorCode, userMessage } } = update;
+
+  if (error) {
+    console.log(errorCode, userMessage);
```
Comment (ivababukova) [inline src/utils/experimentUpdatesHandler.js]: if you want to print the error for debugging, better use `console.error`

File: src/utils/experimentUpdatesHandler.js [hunk 3]
```diff
@@ -63,16 +65,43 @@ const onQCUpdate = (update, dispatch, experimentId) => {
...
 
 const onWorkResponseUpdate = (update, dispatch, experimentId) => {
-  const { request: { body: { name: workRequestName } }, response: { error } } = update;
+  const {
+    request: { body: { name: workRequestName } },
+    response: { error, errorCode, userMessage },
+  } = update;
+
+  if (error) {
+    console.log(errorCode, userMessage);
```
Comment (ivababukova) [inline src/utils/experimentUpdatesHandler.js]: if you want to print the error for debugging, better use `console.error`

**Comment by ivababukova** [comment]
Where am I supposed to see this notification? Because I tried it, the cell set doesn't get created, but I also can't see any update or notification on the UI. From the user perspective, it looks like nothing happened

**Comment by ivababukova** [comment]
Here is a video showing what I mean:

https://user-images.githubusercontent.com/65414651/155681086-34f7fed7-96c2-458d-9589-693503c4823e.mov

File: src/redux/actions/cellSets/createCellSetByExpression.js (hunk 1 above)
Comment (aerlaut) [reply src/redux/actions/cellSets/createCellSetByExpression.js]: This error moved to be logged and handled in the `experimentUpdatesHandler`. We can let the error bubble up to the level above to be handled. If we handle both here and in `experimentUpdatesHandler`, we'll see double error. If we handle it here but not in `experimentUpdatesHandler`, only the user running the createCellSet will see the If we handle it in `experimentUpdatesHandler`, all users currently opening the experiment will see the error.

File: src/utils/experimentUpdatesHandler.js (hunk 2 above)
Comment (aerlaut) [reply src/utils/experimentUpdatesHandler.js]: Oh right, changed

File: src/utils/experimentUpdatesHandler.js (hunk 3 above)
Comment (aerlaut) [reply src/utils/experimentUpdatesHandler.js]: Oh right, changed

**Comment by aerlaut** [comment]
I've update the PR. Can you please try again @ivababukova 

Uploading Screen Recording 2022-02-28 at 10.19.23.mov…

**Comment by aerlaut** [comment]
I've updated the PR. Below is a recording I took with the staging environment. Can you please try again? @ivababukova 

https://user-images.githubusercontent.com/2862362/155966370-f32218f3-4f28-467a-97ae-b0ba5925e297.mov

File: src/redux/actions/cellSets/createCellSetByExpression.js (hunk 1 above)
Comment (ivababukova) [reply src/redux/actions/cellSets/createCellSetByExpression.js]: I don't see the reason why all users should receive a notification for failed cluster creation action. Other users won't know that someone is trying to create a cell set. Getting a failed notification will only confuse them and make them think that they did something that caused the error.

Comment (aerlaut) [reply src/redux/actions/cellSets/createCellSetByExpression.js]: I think we need a proper discussion about this issue, i.e. how we should notify multiple concurrent users viewing the experiment, to be clear on our stance.

Comment (aerlaut) [reply src/redux/actions/cellSets/createCellSetByExpression.js]: Iva has clarified about this during the stand up. Only the user dispatching the work request should be notified, there is no need for others to know about the failed cluster creation. PR has been updated to reflect this.

File: src/utils/work/seekWorkResponse.js [hunk 4]
```diff
@@ -61,11 +61,14 @@ const dispatchWorkRequest = async (
 
   const responsePromise = new Promise((resolve, reject) => {
     io.on(`WorkResponse-${ETag}`, (res) => {
-      const { response: { error } } = res;
+      const { response } = res;
+
+      if (response?.error) {
```
Comment (ivababukova) [inline src/utils/work/seekWorkResponse.js]: Based on my knowledge, there will always be `error` in response

File: src/utils/work/seekWorkResponse.js [hunk 5]
```diff
@@ -61,11 +61,14 @@ const dispatchWorkRequest = async (
...
+        const { errorCode, userMessage } = response;
+        console.error(errorCode, userMessage);
 
-      if (error) {
         return reject(
-          new WorkResponseError(error, request),
+          new WorkResponseError(errorCode, userMessage, request),
```
Comment (ivababukova) [inline src/utils/work/seekWorkResponse.js:71]: will there always be both `userMessage` and `errorCode` in the response when there is an error?

**Comment by ivababukova** [comment]
@aerlaut is the coverage report correct here? I can see lots of tests added in the PR, but the coverage report is all gray, as if no tests are added. Can you verify that the tests you are adding are actually covering code lines and increasing coverage?

File: src/utils/work/seekWorkResponse.js (hunk 4 above)
Comment (aerlaut) [reply src/utils/work/seekWorkResponse.js]: Yes, "error" will always be there for work responses. I'll remove the null check (`?.`)

File: src/utils/work/seekWorkResponse.js (hunk 5 above)
Comment (aerlaut) [reply src/utils/work/seekWorkResponse.js:71]: Yes, there will always be a `userMessage` and `errorCode` when there is an error. The error response is built in `tasks.py` : https://github.com/hms-dbmi-cellenics/worker/blob/master/python/src/worker/tasks/factory.py

File: src/utils/work/seekWorkResponse.js (hunk 4 above)
Comment (aerlaut) [reply src/utils/work/seekWorkResponse.js]: Removed

**Comment by aerlaut** [comment]
> @aerlaut is the coverage report correct here? I can see lots of tests added in the PR, but the coverage report is all gray, as if no tests are added. Can you verify that the tests you are adding are actually covering code lines and increasing coverage?

The tests aren't picked up by code coverage because we only collect coverage from `.jsx` files. Adding `.js` files to the coverage from `.js` files. However, there are some files (e.g. constants, redux "module" `index.js` files whose function is to re-export child functions) that we don't want to collect coverage from. I've added these files and tests under the `coveragePathIgnorePatterns` so that they aren't included in the tests. 

Some of these ignored files (e.g. `pipelineStatusValues`, `endUserMessages`) which contains constants are hard coded into the ignore patterns. However, when there is time, we should group these files together in a folder so the folder can be the one that's ignored. It can be added as a part of this refactor src/utils ticket: https://biomage.atlassian.net/browse/BIOMAGE-1092

**Review by ivababukova — APPROVED** [review]

**Pull request merged**